tests/shadow/test_shadow_class.py

```diff
@@ -0,0 +1,463 @@
+
+
+class TestUnitTestClassicalShadows:
+    """Unit Tests for ClassicalShadow class"""
+
+    @pytest.mark.parametrize("shadow", shadows)
+    def test_unittest_snapshots(self, shadow):
+        """Test the output shape of snapshots method"""
+        T, n = shadow.bits.shape
+        assert (T, n) == shadow.recipes.shape
+        assert shadow.local_snapshots().shape == (T, n, 2, 2)
+        assert shadow.global_snapshots().shape == (T, 2**n, 2**n)
+
+
+class TestIntegrationShadows:
+    """Integration tests for classical shadows class"""
+
+    @pytest.mark.parametrize("shadow", shadows)
+    def test_pauli_string_expval(self, shadow):
+        """Testing the output of expectation values match those of exact evaluation"""
+
+        o1 = qml.PauliX(0)
+        res1 = shadow.expval(o1, k=2)
+
+        o2 = qml.PauliX(0) @ qml.PauliX(1)
+        res2 = shadow.expval(o2, k=2)
+
+        res_exact = 1.0
+        assert qml.math.allclose(res1, res_exact, atol=1e-1)
+        assert qml.math.allclose(res2, res_exact, atol=1e-1)
+
+    Hs = [
+        qml.PauliX(0),
+        qml.PauliX(0) @ qml.PauliX(1),
+        1.0 * qml.PauliX(0),
+        0.5 * qml.PauliX(1) + 0.5 * qml.PauliX(1),
+        qml.Hamiltonian([1.0], [qml.PauliX(0) @ qml.PauliX(1)]),
+    ]
+
+    @pytest.mark.parametrize("H", Hs)
+    @pytest.mark.parametrize("shadow", shadows)
+    def test_expval_input_types(self, shadow, H):
+        """Test ClassicalShadow.expval can handle different inputs"""
+        assert qml.math.allclose(shadow.expval(H, k=2), 1.0, atol=1e-1)
+
+    def test_reconstruct_bell_state(self):
+        """Test that a bell state can be faithfully reconstructed"""
+        wires = range(2)
+
+        dev = qml.device("default.qubit", wires=wires, shots=10000)
+
+        @qml.qnode(dev)
+        def qnode(n_wires):
+            qml.Hadamard(0)
+            qml.CNOT(wires=[0, 1])
+            return qml.classical_shadow(wires=range(n_wires))
+
+        # should prepare the bell state
+        bits, recipes = qnode(2)
+        shadow = ClassicalShadow(bits, recipes)
+        global_snapshots = shadow.global_snapshots()
+
+        state = np.sum(global_snapshots, axis=0) / shadow.snapshots
+        bell_state = np.array([[0.5, 0, 0, 0.5], [0, 0, 0, 0], [0, 0, 0, 0], [0.5, 0, 0, 0.5]])
+        assert qml.math.allclose(state, bell_state, atol=1e-1)
+
+        # reduced state should yield maximally mixed state
+        local_snapshots = shadow.local_snapshots(wires=[0])
+        assert qml.math.allclose(np.mean(local_snapshots, axis=0)[0], 0.5 * np.eye(2), atol=1e-1)
+
+        # alternative computation
+        bits, recipes = qnode(1)
+        shadow = ClassicalShadow(bits, recipes)
+        global_snapshots = shadow.global_snapshots()
+        local_snapshots = shadow.local_snapshots(wires=[0])
+
+        state = np.sum(global_snapshots, axis=0) / shadow.snapshots
+        assert qml.math.allclose(state, 0.5 * np.eye(2), atol=1e-1)
```

Comment:
Maybe you could test that in the latter case the global and local snapshots are the same (removing the unnecessary dimension).
```suggestion
        assert qml.math.allclose(state, 0.5 * np.eye(2), atol=1e-1)
        assert np.all(local_snapshots[:, 0] == global_snapshots)
```

Reply:
Tagging @Qottmann for this since he wrote the test, and I'm not exactly sure if the local and global snapshots can be compared this way.

Reply:
Good idea! @eddddddy it should be the same in that case indeed because I am only looking at a single site